src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,432 @@
+    ReportTimestamp = erlang:system_time(millisecond),
+    true = ets:insert(?ETS, {ReportTimestamp, EncodedPacket}),
+    ok.
+
+-spec write_packets() -> ok.
+write_packets() ->
+    gen_server:cast(?SERVER, write).
+
+-spec upload_packets(FilePath :: string()) -> ok.
+upload_packets(FilePath) ->
+    gen_server:cast(?SERVER, {upload, FilePath}).
+
+-spec restart_report_interval(IntervalDuration :: interval_duration_ms()) -> ok.
+restart_report_interval(IntervalDuration) ->
+    gen_server:cast(?SERVER, {start_interval, IntervalDuration}).
+
+-spec stop_report_interval() -> ok.
+stop_report_interval() ->
+    gen_server:cast(?SERVER, stop_interval).
+
+%%%===================================================================
+%%% gen_server Function Definitions
+%%%===================================================================
+
+init(
+    #{
+        write_dir := WriteDir,
+        max_file_size := MaxFileSize,
+        interval_duration := IntervalDuration,
+        upload_window := UploadWindow,
+        bucket := Bucket
+    } = Args
+) ->
+    AWSClient = setup_aws(Args),
+    ok = filelib:ensure_dir(WriteDir),
+    TempFilePath = generate_file_name(WriteDir),
+    {ok, ReportInterval} = start_report_interval(IntervalDuration),
+    {ok, #state{
+        aws_client = AWSClient,
+        write_dir = WriteDir,
+        file_path = TempFilePath,
+        max_file_size = MaxFileSize,
+        report_interval = ReportInterval,
+        interval_duration = IntervalDuration,
+        upload_window = UploadWindow,
+        upload_window_start_time = erlang:system_time(millisecond),
+        bucket = Bucket
+    }}.
+
+handle_call(Msg, _From, State = #state{}) ->
+    {stop, {unimplemented_call, Msg}, State}.
+
+handle_cast(
+    write,
+    State = #state{
+        write_dir = WriteDir,
+        file_path = FilePath,
+        max_file_size = MaxFileSize,
+        upload_window_start_time = WindowStartTime,
+        upload_window = UploadWindow
+    }
+) ->
+    {ok, FileSize} = handle_write(FilePath, MaxFileSize, ?FILE_WRITE_OPTIONS),
+
+    case FileSize >= MaxFileSize orelse upload_window_elapsed(WindowStartTime, UploadWindow) of
+        true ->
+            ?MODULE:upload_packets(FilePath),
+            {noreply, State#state{file_path = generate_file_name(WriteDir)}};
+        false ->
+            {noreply, State}
+    end;
+handle_cast({upload, FilePath}, State = #state{aws_client = AWSClient, bucket = Bucket}) ->
+    UploadTimestamp = erlang:system_time(millisecond),
+    FileName = list_to_binary("packetreport." ++ integer_to_list(UploadTimestamp) ++ ".gz"),
+
+    case upload_file(AWSClient, list_to_binary(FilePath), FileName, Bucket) of
+        {ok, _} ->
+            file:delete(FilePath);
+        Error ->
+            lager:warning("packet reporter failed to upload: ~p~n", [Error]),
+            error
+    end,
+    {noreply, State#state{upload_window_start_time = UploadTimestamp}};
```

Comment:
If the upload fails, should we still advance the upload window start time? I'm thinking we should not so the next time we check we'll try again, rather than waiting an entire interval.

Reply:
This could be an option. 
Upload failures are retried up to five times with a minute between retries. Per conversations with Madninja, the packets may be sufficiently stale after that point that reporting them wouldn't be a priority.

When upload is called the write file is rotated, so calling upload on the next interval (as is) would upload the new write file. I could see incorporating some kind of retry queue as an improvement.